### ISSUE TITLE
Strange universe inconsistencies

### ISSUE THREAD

**Issue by Ekdohibs** [coq/coq#7875]
#### Version

The Coq Proof Assistant, version 8.7.2 (March 2018)
compiled on Mar 19 2018 11:03:14 with OCaml 4.06.1

#### Operating system

Linux 4.13.0-45-generic #50~16.04.1-Ubuntu

#### Description of the problem

When using the [VPL](https://github.com/VERIMAG-Polyhedra/VPL), the following code produces an universe inconsistency:
```
Require Import ImpureConfig.
Require Import PedraQ.
Require Import List.
Definition f (x : unit) : imp unit := pure x.
Definition g l := map f l.
```

As noticed  by @boulme on the coq-club mailing list, this seems somehow specific to `map`: a variant with `fold_right (fun x l => (f x) :: l) nil l` does not trigger the inconsistency, nor does `fold_left`.

Now, go to the file `Map_poly.v` of the VPL and find the following line: `rewrite H1 in H0 at 1`.

First case: replace it by: ```
    Print Universes "a".

Definition f (x : unit) : imp unit := pure x.
Definition g l := map (fun x => f x) l.

    Print Universes "b".
    rewrite H1 in H0 at 1.
    Print Universes "c".

Definition f1 (x : unit) : imp unit := pure x.
Definition g1 l := map (fun x => f1 x) l.
```
The definitions of `f`, `g`, `f1`, `g1` are all accepted.

Second case: replace it by: ```

    Print Universes "d".
    rewrite H1 in H0 at 1.
    Print Universes "e".

Definition f (x : unit) : imp unit := pure x.
Definition g l := map (fun x => f x) l.
```
This time, the definition of `g` is rejected with an universe inconsistency message.

Finally, perhaps the most surprising:
```
$ diff -s b c
Files b and c are identical
$ diff -s d e
Files c and d are identical
$ diff -s a d
Files a and d are identical
$ diff -s b d
2638d2637
< Vpl.Impure.2 <= Coq.Lists.List.170 ;
```
That is, the definition of `f` and `g` is accepted in the first case, but not in the second, while the universes that are printed are the same.

**Comment by JasonGross** [comment]
If you want to make a minimal self-contained test case, you can use my ` find-bug.py` script at https://github.com/JasonGross/coq-tools.  This would probably help with debugging.

Random guesses: are there any places where you do `map list <some list of types>` , or ` Definition <some name> := @eq.` or ` Fixpoint <something> := list <something with a recursive call>`?  All of these can capture template universes.  (I guess if you have a `Proper` instance for `list` or `@eq`, that can also capture template universes.)

**Comment by JasonGross** [comment]
Also, if you pass `list` or `option` ( rather than ` fun T => list T` or `fun T => option T` as, e.g., a monad to a function that is parametric over monads ` Type -> Type`, that can capture template universes.

**Comment by Ekdohibs** [comment]
With help from `find-bug.py`, I was able to reduce the test case to the following:
```
Require Import Morphisms.
Require Import List.

Module Type ImpureMonad.

  Axiom imp: Type -> Type.
  Axiom pure: forall {A}, A -> imp A.
  Axiom mayReturn: forall {A : Type}, imp A -> A -> Prop.

End ImpureMonad.

Module PureImpureMonad <: ImpureMonad.

   Definition imp (A : Type) := A.

   Definition mayReturn {A : Type} (a b : A) : Prop := a = b.

   Definition pure {A : Type} (a : A) := a.

   Instance mayReturn_eq_compat A: Proper (@eq A ==> @eq A ==> iff) (@mayReturn A).
   Proof.
     intros a b H1 c d H2. rewrite H1, H2. tauto.
   Qed.

End PureImpureMonad.

Module ForceConstraint : ImpureMonad := PureImpureMonad.

Lemma dummy_using_setoid: forall p e: nat, e = p -> p = e.
Proof.
  intros p e H.
  assert (H1 : e = e) by reflexivity.
  rewrite H in H1 at 1.
  exact H1.
Qed.

Definition f (x : unit) : PureImpureMonad.imp unit := PureImpureMonad.pure x.
Definition g l := map (fun x => f x) l.
```
Again, if the definitions of `f` and `g` are added before the `dummpy_using_setoid` lemma, there is no error, and the same definitions can be put afterwards without error.

**Comment by eponier** [comment]
What is funny is that if the proof of `dummy_using_setoid` is processed asynchronously (with "go to end" or "go to cursor" in CoqIDE), the definition of `g` does not produce any error, while if it is processed synchronously, the error shows up.

**Comment by ppedrot** [comment]
@eponier You have to click on the gears button to force the computation of constraints, that will raise the error most probably.

**Comment by eponier** [comment]
@ppedrot Right, I get an error, but probably not the one that you expected:
`Anomaly: Uncaught exception Univ.UniverseInconsistency(_). Please report at http://coq.inria.fr/bugs/.`

**Comment by eponier** [comment]
By the way, I also observed a strange behavior of CoqIDE with respect to the "View" menu that I should probably report in another issue. If I enable "display universe levels", then enable and afterwards disable "display all low-level contents", then universe levels are no longer displayed while "display universe levels" is still shown as enabled.

**Comment by vbgl** [comment]
For the record, in recent master, the error message is slightly more informative:

> Error:
In environment
l : ?T
x : unit
The term "f x" has type "PureImpureMonad.imp unit" while it is expected to have type "?B"
(unable to find a well-typed instantiation for "?B": cannot ensure that "Type@{PureImpureMonad.imp.u0}" is a
subtype of "Type@{Coq.Lists.List.170}").

**Comment by SkySkimmer** [comment]
>~~~
>$ diff -s b c
>Files b and c are identical
>$ diff -s d e
>Files c and d are identical
>$ diff -s a d
>Files a and d are identical
>$ diff -s b d
>2638d2637
>< Vpl.Impure.2 <= Coq.Lists.List.170 ;
>~~~

Not sure what happened here. Especially why `diff -s d e` talks about files `c` and `d`.

**Comment by Ekdohibs** [comment]
I think the `diff -s d e` talking about files `c` and `d` is me changing the name of the files after-the-fact and forgetting to rename all uses consistently (it should be `d` and `e` being identical). The important point is rather that the universes printed before and after the proof with setoid are both equal, but that the presence of this proof raises the inconsistent universe problem, while it does not modify universes constraints at all!

**Comment by SkySkimmer** [comment]
So it's `a = d = e` and `b = c`?

Note that Print Universes inside a proof actually prints the global universes unaffected by the proof in progress.

**Comment by Ekdohibs** [comment]
Yes, these are the correct equalities. I didn't know that about universes inside a proof. It seems like the universes are indeed different before and after the proof of `dummy_using_setoid` (in the simplified test case).
The inconsistency that is produced is still strange, since `Module ForceConstraint : ImpureMonad := PureImpureMonad.` seems somehow necessary to trigger the inconsistency.

**Comment by SkySkimmer** [comment]
No inconsistency in the minimized version (didn't test the original) now. Possibly some template constraint about the univ of eq which now doesn't get generated?
